src/org/sosy_lab/java_smt/api/UserPropagator.java

```diff
@@ -0,0 +1,114 @@
+ * rather than implementing this interface directly.
+ *
+ * <p> A user propagator provides various callbacks that are invoked by the solver during the
+ * solving process. Within these callbacks, the user can react to and influence the solver
+ * by calling into the {@link PropagatorBackend}. For example, he can raise conflicts
+ * whenever the solver makes assignments inconsistent to the user-defined theory.
+ */
+public interface UserPropagator {
+
+  /**
+   * This callback is invoked whenever the solver creates a new decision level.
+   * A user propagator should maintain backtracking points on each push to enable later
+   * backtracking.
+   */
+  void onPush();
+
+  /**
+   * This callback is invoked when the solver backtracks. The solver can backtrack multiple
+   * levels simultaneously.
+   * @param numPoppedLevels The number of levels to backtrack (~ number of pushes to backtrack).
+   */
+  void onPop(int numPoppedLevels);
+
+  /**
+   * This callback is invoked when the solver finds a complete satisfying assignment.
+   * The user can check the found model for consistency and potentially raise conflicts via
+   * {@link PropagatorBackend#propagateConflict(BooleanFormula[])}.
+   * <p>
+   * Note: This callback is only invoked if the user propagator enabled it via
+   * {@link PropagatorBackend#notifyOnFinalCheck()}.
+   */
+  void onFinalCheck();
+
+  /**
+   * This callback is invoked if the solver gets to know the value of a registered expression
+   * ({@link #registerExpression}).
+   * Within the callback, the user can raise conflicts via
+   * {@link PropagatorBackend#propagateConflict}, propagate consequences via
+   * {@link PropagatorBackend#propagateConsequence}, or influence the solvers decision heuristics
+   * via {@link PropagatorBackend#propagateNextDecision}.
+   * <p>
+   * The reported value is only known on the current and later push levels,
+   * but will get invalidated when backtracking.
+   * <p>
+   * Note: This callback is only invoked if the user propagator enabled it via
+   * {@link PropagatorBackend#notifyOnKnownValue()}.
+   *
+   * @param expr  The expressions whose value is known.
+   * @param value The value of the expression.
+   */
+  void onKnownValue(BooleanFormula expr, boolean value);
+
+  /**
+   * This callback is invoked if the solver decides to branch on a registered expression.
+   * ({@link #registerExpression}).
+   * Within the callback, the user can change the decision by calling
+   * {@link PropagatorBackend#propagateNextDecision}.
+   * <p>
+   * Note: This callback is only invoked if the user propagator enabled it via
+   * {@link PropagatorBackend#notifyOnDecision()}.
+   * @param expr The expressions whose value gets decided (usually a literal).
+   * @param value The decision value.
+   */
+  void onDecision(BooleanFormula expr, boolean value);
+
+  /**
+   * Connects this user propagator with a {@link PropagatorBackend}. The backend is used
```

Comment:
Can this method be called mutliple times?
If yes, is any existing propagator replaced by the new one?
If no, do we want to throw an exception?
PLease document this behaviour.

Reply:
Can we add a user-propagator at each point in time, e.g., after pushing several formulas? Or should we create a user propagator before creating any symbols? For Z3, it seems to be sufficient to just instantiate the uSerPropagator right before callung the SMT/SAT-check.

Reply:
I have not tried out what the limitations are, but it is definitely fine to add the user propagator anytime before the first SAT-check. You can create the whole solver formula and then add the propagator (you need to register the expressions after adding the propagator though)
Whether one can add them after the first SAT-check is something I didn't test.

Reply:
After thinking more about this, I think it is always bad to register the same user propagator multiple times.
I made this point clear in the documentation and the implementation provided by `AbstractUserPropagator` will throw an error if the method is called multiple times.